File: c/vector.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

/*
 * Implement a vector (mutable array with automatic resizing):
*/

typedef struct Vector {
    int     size;
    int     capacity;
    int*    container;
    int     (*get_size)(struct Vector*);
    int     (*get_capacity)(struct Vector*);
    bool    (*is_empty)(struct Vector*);
    int     (*at)(struct Vector*, int);
    void    (*resize)(struct Vector*);
    void    (*append)(struct Vector*, int);
    void    (*prepend)(struct Vector*, int);
    void    (*insert)(struct Vector*, int, int);
    int     (*pop)(struct Vector*);
    void    (*delete)(struct Vector*, int);
    void    (*remove)(struct Vector*, int);
    int     (*find)(struct Vector*, int);
}   Vector;

int get_size(Vector* self)
{
    return self->size;
}

int get_capacity(Vector* self)
{
    return self->capacity;
}

bool is_empty(Vector* self)
{
    if (self->size == 0)
        return true;
    return false;
}

int at(Vector* self, int index)
{
    if (index >= 0 && index < self->size)
        return self->container[index];
    return -1;
}

void resize(Vector* self)
{
    int* newContainer = NULL;

    if (self == NULL)
        return ;

    self->capacity *= 2;
    newContainer = calloc(self->capacity, sizeof(int));
    for (int offset = 0; offset < self->size; offset++)
    {
        *(newContainer + offset) = *(self->container + offset);
    }
    free(self->container);
    self->container = newContainer;
}

void append(Vector* self, int num)
{
    if (self->size == self->capacity)
    {
        self->resize(self); // O(2*N)
    }
    *(self->container + self->size) = num; // O(1)
    self->size++; // O(1)
}

void insert(Vector* self, int num, int idx)
{
    if (idx < 0) idx = 0;
    if (idx > self->size)
    {
        self->append(self, num);
        return ;
    }
    if (self->size == self->capacity) self->resize(self);

    for (int offset = self->size-1; offset >= idx; offset--)
    {
        *(self->container + offset + 1) = *(self->container + offset);
        if (offset == idx)
        {
            *(self->container + offset) = num;
            break ;
        }
    }
    self->size++;
}

void prepend(Vector* self, int num)
{
    self->insert(self, num, 0);
}

// remove from end, return value
int pop(Vector* self)
{
    if (self->is_empty(self))
        return -1;
    
    self->size--;
    return *(self->container + self->size);
}

// delete item at index, shifting all trailing elements left
void delete(Vector* self, int idx)
{
    if (self->is_empty(self) || idx < 0 || idx >= self->size)
        return ;

    while (idx < self->size-1)
    {
        *(self->container + idx) = *(self->container + idx + 1);
        idx++;
    }
    self->size--;
}

// looks for value and removes index holding it (even if in multiple places)
void remove_fn(Vector* self, int target)
{
    int count = 0, insert_pos = 0;
    int* newContainer = NULL;

    // count
    for (int offset = 0; offset < self->size; offset++)
    {
        if (*(self->container + offset) == target)
            count++;
    }
    // allocate
    newContainer = calloc(self->size - count, sizeof(int));
    if (newContainer == NULL)
        return ;
    // copy
    for (int offset = 0; offset < self->size; offset++)
    {
        if (*(self->container + offset) != target)
            *(newContainer + insert_pos) = *(self->container + offset);
    }
    free(self->container);
    self->container = newContainer;
}

// looks for value and returns first index with that value, -1 if not found
int find(Vector* self, int target)
{
    if (self == NULL)
        return -1;

    for (int offset = 0; offset < self->size; offset++)
    {
        if (*(self->container + offset) == target)
            return offset;
    }
    return -1;
}

Vector* newVector()
{
    Vector* v = calloc(1, sizeof(Vector));

    v->size = 0;
    v->capacity = 4;
    v->container = calloc(v->capacity, sizeof(int));
    v->get_size = get_size;
    v->get_capacity = get_capacity;
    v->is_empty = is_empty;
    v->at = at;
    v->append = append;
    v->prepend = prepend;
    v->insert = insert;
    v->resize = resize;
    v->pop = pop;
    v->remove = remove_fn;
    v->delete = delete;
    v->find = find;
    return v;
}

int main()
{
    Vector* v = newVector();
    
    printf("Size: %d\n", v->get_size(v));
    printf("Capacity: %d\n", v->get_capacity(v));

    v->append(v, 45);
    v->append(v, 65);
    v->append(v, 100);
    v->append(v, 9);
    v->append(v, 999);
    v->insert(v, 1188, 40);

    for (int i = 0; i < v->get_size(v); i++)
    {
        printf("[%d] %d\n", i, *(v->container + i));
    }

    printf("Size: %d\n", v->get_size(v));
    printf("Capacity: %d\n", v->get_capacity(v));

    free(v->container); free(v);
    return 0;
}

File: src/data_structures/_vector.py
from typing import List

class Vector:
    def __init__(self, items: List[int]=[]) -> None:
        self._size = len(items)
        self._capacity = (len(items)+1) * 2
        self._container = [0] * self._capacity
        for i, x in enumerate(items):
            self._container[i] = x
    
    # occupied space
    def size(self) -> int:
        return self._size

    # allocated space
    def capacity(self) -> int:
        return self._capacity

    # check if vector is empty
    def is_empty(self) -> bool:
        if self.size == 0:
            return True
        return False

    # return num at specific index
    def at(self, idx: int) -> int:
        if idx < 0 or idx >= self._size:
            return None
        return self._container[idx]

    # increase/decrease capacity based on size
    def _resize(self) -> None:
        if self._size == self._capacity:
            self._capacity *= 2
        elif self._size <= self._capacity // 4:
            self._capacity /= 2
        
        newArr = [0] * self._capacity
        for i in range(self._size):
            newArr[i] = self._container[i]
        self._container = newArr
    
    # insert at end
    def append(self, num: int) -> None:
        if self._size == self._capacity:
            self._resize()
        self._container[self._size] = num
        self._size += 1

    # insert at begining
    def prepend(self, num: int) -> None:
        self.insert(num, 0)

    # insert num at specific index
    def insert(self, num: int, idx: int) -> None:
        if idx > self._size:
            return
        if self._size == self._capacity:
            self._resize()

        for i in range(self._size, idx-1, -1):
            if i == idx:
                self._container[i] = num
                break
            self._container[i] = self._container[i-1]
        self._size += 1

    # remove and return last num
    def pop(self):
        if self._size > 0:
            self._size -= 1
            return self._container[self._size]
        return None

    # delete item at index, shifting all trailing elements left
    def delete(self, index: int):
        if self.is_empty():
            return
    
        for i in range(index, len(self._container)-1):
            self._container[i] = self._container[i+1]
        self._size -= 1

    # // looks for value and removes index holding it (even if in multiple places)
    def remove(self, target: int) -> None:
        new_size = self._size - self._container.count(target)
        tmp, pos = [0] * self._capacity, 0


        for i in range(self._size):
            if self._container[i] != target:
                tmp[pos] = self._container[i]
                pos += 1
        self._container = tmp

        self._size = pos

        """ or
            v = self._container
            tmp = [v[i] for i in range(self._size) if v[i] != target]
            self._container = v
            self._size = len(tmp)
        """
        
    # looks for value and returns first index with that value, -1 if not found
    def find(self, num: int) -> int:
        for i, n in enumerate(self._container):
            if n == num:
                return i

    def __repr__(self) -> str:
        return f'Vector of size: {self._size} and capacity {self._capacity}:\n{self._container[:self._size]}'

if __name__ == '__main__':
    v = Vector([1, 2, 3])
    v.append(4)
    v.append(6)
    v.insert(5, 4)
    v.pop()
    v.insert(6, 5)
    v.append(2)
    v.append(2)
    v.append(7)
    v.remove(2)
    v.insert(2, 1)
    v.find(3)
    v.delete(0)
    print(f'capacity: {v.capacity()}')
    print(f'size: {v.size()}')
    print(v)
    
